Template Pipeline: Null-Safety

dylhunn opened the pull request:
Various improvements to the template pipeline, including

- null-safe navigation (both property and keyed reads)
- non-null assertions
- initial processing of nullish coalescing

Several features, such as reads of pipes or null-safe calls, will require generating temporary variables to be fully complete. This will be handled in a temporary variables pass, in a subsequent PR.

See individual commits for details.

dylhunn committed:
refactor(compiler): Allow expressions to be deeply cloned.

It is sometimes useful to clone an expression tree, in order to copy it and mutate it in a phase, without affecting other subtrees due to the copy-by-reference.

dylhunn committed:
refactor(compiler): Add initial support for nullish coalescing.

Create a pass that expands nullish coalescing operators into null checks.

This is not yet finished because we need to emit temporary variable assignments, which we will do using a future temporary variable allocation pass. Also, TemplateDefinitionBuilder is a bit quirky, and we still need to exactly match its behavior.

Nevertheless, this is good enough to prevent the diffs from getting ruined as a result of nullish coalescing operations.

mmalerba commented on packages/compiler/src/template/pipeline/src/phases/nullish_coalescing.ts:
```diff
@@ -0,0 +1,38 @@
+          return expr;
+        }
+
+        // TODO: We need to unconditionally emit a temporary variable to match
```
Should we add a compatibility flag to this phase too? Or is this comment just a reminder to look into optimizing it in the future?

dylhunn replied:
This is not only a reminder that a compatibility-breaking optimization could be performed in the future -- it's also *wrong* as proposed today. We need to emit that temporary to have the correct behavior.

Stay tuned, I have a temporary variables pass on deck!

mmalerba commented on packages/compiler/src/template/pipeline/src/phases/expand_safe_reads.ts:
```diff
@@ -0,0 +1,94 @@
+  }
+
+  if (e instanceof ir.SafeInvokeFunctionExpr) {
+    // TODO
```
Are you intentionally saving this for a followup PR?

dylhunn replied:
Yes. Don't want to bite off more than I can chew -- function calls add extra "layers" to the AST and it's a huge pain to deal with.

alxhub commented on packages/compiler/src/template/pipeline/src/phases/register_element_attributes.ts:
```diff
@@ -0,0 +1,68 @@
+import * as o from '../../../../output/output_ast';
+import {Identifiers as R3} from '../../../../render3/r3_identifiers';
+import * as ir from '../../ir';
+import {ComponentCompilation, ViewCompilation} from '../compilation';
```
```suggestion
import type {ComponentCompilation, ViewCompilation} from '../compilation';
```

alxhub commented on packages/compiler/src/template/pipeline/src/phases/register_element_attributes.ts:
```diff
@@ -0,0 +1,68 @@
+/**
+ * Find all RegisterAttribute ops, and collect them into the ElementAttribute structures.
+ */
+export function phaseAttributeExtraction(cpl: ComponentCompilation, compatibility: boolean): void {
```
I think we should actually put it as a flag on `ComponentCompilation` since it's used across multiple phases now, and document it that way. We can do that in a followup.

dylhunn replied:
SGTM. I'll send a refactoring commit in the next PR.

alxhub commented on packages/compiler/src/template/pipeline/src/phases/register_element_attributes.ts:
```diff
@@ -0,0 +1,68 @@
+        ir.assertIsElementAttributes(ownerOp.attributes);
+
+        let extractable = compatibility ?
+            (op.value instanceof o.LiteralExpr && typeof op.value.value === 'string') :
```
Maybe a comment here on what's considered "extractable" in both modes? Also `extractable` can be `const` I think.

alxhub commented on packages/compiler/src/template/pipeline/src/phases/nullish_coalescing.ts:
```diff
@@ -0,0 +1,38 @@
+
+import * as o from '../../../../output/output_ast';
+import * as ir from '../../ir';
+import {ComponentCompilation} from '../compilation';
```
```suggestion
import type {ComponentCompilation} from '../compilation';
```

alxhub reviewed (approved):
LGTM pending Miles' comments as well as mine.

dylhunn commented:
caretaker: this PR has no impact on g3 and is safe to merge.

dylhunn committed:
refactor(compiler): Support non-null assertions.

Because non-null assertions don't affect the generated code, we can just drop them.

dylhunn committed:
refactor(compiler): Support safe property reads and keyed reads.
Angular's null-safe access operators differ from Javascript's built-in semantics, in that they short-circuit to `null` instead of `undefined`. This necessitates providing a custom transformation, instead of relying on Typescript or Javascript itseld.

The old TemplateDefinitionBuilder uses a top-down approach based on the Visitor pattern, in which it recursively extracts the left-most safe access, and hoists it to a null check at the top. See `expression_converter.ts` for details.

In this commit, we replace that approach with a new bottom-up algorithm, as part of the template pipeline. This requires an intermediate expression type to represent the not-yet-expanded ternary operators, and is split into its own pass.

Null-safe function calls are not yet implemented, since they will rely on a future temporary variable allocation pass.

Co-authored-by: Alex Rickabaugh <alxhub@users.noreply.github.com>

pkozlowski-opensource commented:
@dylhunn it looks like G3 presubmit wasn't run on this PR (?) - not sure what is going on but we should have G3 sanity check before merging this one.

dylhunn commented:
Presubmit link is http://test/OCL:539216144:BASE:539216167:1686357100639:9bb91c78
